Enhance development documentation regarding parallel testing (#2336)
SAP/jenkins-library: DEVELOPMENT.md

(See [blog about it](https://eleni.blog/2019/05/11/parallel-test-execution-in-go/).)
At the minimum, you need to capture the test case value from the loop iteration variable, by shadowing this variable in the loop body.
Inside the `t.Run()` closure, this shadow copy is visible, and cannot be overwritten by later loop iterations.
If you do not make this shadowing copy, what is visible in the closure is the variable which gets re-assigned with a new value in each loop iteration.
The value of this variable is then not fixed for the test run.

```go
func TestMethod(t *testing.T) {
    t.Parallel() // indicates that this method can parallel to other methods
    testCases := []struct {
        Name string
    }{
        {
            Name: "Name1"
        },
        {
            Name: "Name2"
        },
    }

    for _, testCase := range testCases { // testCase defined here is re-assigned in each iteration
        testCase := testCase // define new variable within loop to detach from overwriting of the outer testCase variable by next loop iteration
        // The same variable name "testCase" is used for convenience.

>         t.Run(testCase.Name, func(t *testing.T){

Reviewer: ```suggestion
        t.Run(testCase.Name, func(t *testing.T) {
```

... go format

### Test Parallelization

Tests that can be executed in parallel should be marked as such. With the command `t.Parallel()` the test framework can
be notified that this test can run in parallel, and it can start running the next test.
([Example in Stackoverflow](https://stackoverflow.com/questions/44325232/are-tests-executed-in-parallel-in-go-or-one-by-one))
Therefore, this command shall be called at the beginning of a test method and `t.Run()` sub tests.
See also the [documentation](https://golang.org/pkg/testing/#T.Parallel) for `t.Parallel()` and `t.Run()`.

```go
func TestMethod(t *testing.T) {
    t.Parallel() // indicates that this method can parallel to other methods

    t.Run("sub test 1", func(t *testing.T){
        t.Parallel() // indicates that this sub test can run parallel to other sub tests
        // execute test
    })

    t.Run("sub test 2", func(t *testing.T){
        t.Parallel() // indicates that this sub test can run parallel to other sub tests
        // execute test
    })
}
```

Go will first execute the non-parallelized tests in sequence and afterwards execute the all parallel tests in parallel
limited by the default number of parallel executions.

> For table tests some additional lines are necessary to prevent overshadowing of the iteration variable.

Reviewer: ```suggestion
It is important that tests executed in parallel use the variable values actually meant to be visible to them.
Especially in table tests, it can happen easily that a variable injected into the `t.Run()`-closure via the outer scope is changed before or while the closure executes.
To prevent this, it is possible to create shadowing instances of variables in the body of the test loop.
```

Author: Sounds better :-)

(See [blog about it](https://eleni.blog/2019/05/11/parallel-test-execution-in-go/).)

> Additionally, to the `t.Parallel()` a capture of the iteration variable is necessary.

Reviewer: ```suggestion
At the minimum, you need to capture the test case value from the loop iteration variable, by shadowing this variable in the loop body.
Inside the `t.Run()` closure, this shadow copy is visible, and cannot be overwritten by later loop iterations.
If you do not make this showing copy, what is visible in the closure is the variable which gets re-assigned with a new value in each loop iteration.
The value of this variable is then not fixed for the test run.
```

> The loop opens with each iteration a new scope. The used variable in the loop is only defined outside the scope
> such that each loop overwrites the variable which will lead to execution of loops with the same variable multiple times
> and leave out some others. Therefore, the variable shall be captured by defining a new variable within the loops scope
> that represents the looping variable.

Reviewer: ```suggestion
```

```go
func TestMethod(t *testing.T) {
    t.Parallel() // indicates that this method can parallel to other methods
    testCases := []struct {
        Name string
    }{
        {
            Name: "Name1"
        },
        {
            Name: "Name2"
        },
    }

>     for i, name := range testCases { // name is only defined outside the loop
>         name := name // define new variable within loop to prevent overwrite of name variable outside loop
>         // Otherwise the test with "Name2" could get executed twice and "Name1" never.
>         // The same variable name "name" is used for convinience.
>         t.Run("sub test " + i, func(t *testing.T){
>             t.Parallel() // indicates that this sub test can run parallel to other sub tests
>             // execute test
>         })
>     }

Reviewer: ```suggestion
    for _, testCase := range testCases { // testCase defined here is re-assigned in each iteration
        testCase := testCase // define new variable within loop to detach from overwriting of the outer testCase variable by next loop iteration
        // The same variable name "testCase" is used for convenience.
        t.Run(testCase.Name, func(t *testing.T){
            t.Parallel() // indicates that this sub test can run parallel to other sub tests
            // execute test
        })
    }
```
